src/main/java/com/amazon/ion/impl/IonReaderContinuableCoreBinary.java

```diff
@@ -1369,10 +1378,301 @@ private enum State {
+                    }
+                    if (parameterPresence == PresenceBitmap.GROUP) {
+                        readGroupExpression(encoding, expressions);
+                    } else {
+                        throw new IllegalStateException();
+                    }
+                    break;
+                case ZeroOrMore:
+                    if (parameterPresence == PresenceBitmap.EXPRESSION) {
+                        readSingleExpression(encoding, expressions);
+                    } else if (parameterPresence == PresenceBitmap.GROUP) {
+                        readGroupExpression(encoding, expressions);
+                    } else if (parameterPresence != PresenceBitmap.VOID) {
+                        throw new IllegalStateException();
+                    }
+                    break;
+            }
+        }
+
+        /**
+         * Collects the expressions that compose the current macro invocation.
+         * @param expressions receives the expressions as they are materialized.
+         */
+        private void collectMacroInvocationExpressions(List<Expression.EExpressionBodyExpression> expressions) {
+            if (isSystemInvocation()) {
+                throw new UnsupportedOperationException("System macro invocations not yet supported.");
+            }
+            int invocationStartIndex = expressions.size();
+            long id = getMacroInvocationId();
+            MacroRef address = MacroRef.byId(id);
+            Macro macro = macroEvaluator.getEncodingContext().getMacroTable().get(address);
+            PresenceBitmap presenceBitmap = new PresenceBitmap();
+            List<Macro.Parameter> signature = macro.getSignature();
+            presenceBitmap.initialize(signature);
+            if (presenceBitmap.getByteSize() > 0) {
+                if (fillArgumentEncodingBitmap(presenceBitmap.getByteSize()) == Event.NEEDS_DATA) {
+                    throw new UnsupportedOperationException("TODO: support continuable parsing of AEBs.");
+                }
+                presenceBitmap.readFrom(buffer, (int) valueMarker.startIndex);
+                presenceBitmap.validate();
+            }
+            expressions.add(Expression.Placeholder.INSTANCE);
+            int numberOfParameters = presenceBitmap.getTotalParameterCount();
+            for (int i = 0; i < numberOfParameters; i++) {
+                readParameter(signature.get(i), presenceBitmap.get(i), expressions);
+            }
+            expressions.set(invocationStartIndex, new Expression.EExpression(address, invocationStartIndex, expressions.size()));
+        }
+
+        /**
+         * Materializes the expressions that compose the macro invocation on which the reader is positioned and feeds
+         * them to the macro evaluator.
+         */
+        private void beginEvaluatingMacroInvocation() {
+            List<Expression.EExpressionBodyExpression> expressions = new ArrayList<>();
```

Comment:
In the future, it might help to have `macroEvaluator` maintain a pool of these lists to re-use.

Reply:
Added a TODO comment about this.